File: spring-backend/src/main/java/com/petcare/service/AuthService.java
package com.petcare.service;

import com.petcare.model.Usuario;
import com.petcare.repository.UsuarioRepository;
import com.petcare.security.JwtTokenService;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;
import java.util.Optional;

@Service
public class AuthService {
    private final UsuarioRepository usuarioRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenService jwtTokenService;

    public AuthService(UsuarioRepository usuarioRepository, PasswordEncoder passwordEncoder, JwtTokenService jwtTokenService) {
        this.usuarioRepository = usuarioRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenService = jwtTokenService;
    }

    public Optional<Usuario> findByUsername(String username) {
        return usuarioRepository.findByUsername(username);
    }

    public Optional<String> login(String username, String password) {
        return usuarioRepository.findByUsername(username)
            .filter(user -> passwordEncoder.matches(password, user.getPasswordHash()))
            .map(user -> jwtTokenService.createToken(user.getId(), user.getUsername(), user.getEmail(), user.getRol()));
    }

    public Usuario createUser(String username, String email, String password, String rol) {
        Usuario user = new Usuario();
        user.setUsername(username);
        user.setEmail(email);
        user.setPasswordHash(passwordEncoder.encode(password));
        user.setRol(rol == null || rol.isBlank() ? "gestor" : rol);
        user.setIsActive(true);
        return usuarioRepository.save(user);
    }

    public Optional<String> recoverPassword(String email) {
        return usuarioRepository.findByEmail(email).map(user -> {
            user.setResetToken(UUID.randomUUID().toString());
            user.setResetTokenExpires(OffsetDateTime.now().plusHours(1));
            usuarioRepository.save(user);
            return user.getResetToken();
        });
    }
}

File: src/main/kotlin/com/petcare/service/AuthService.kt
package com.petcare.service

import com.petcare.model.Session
import com.petcare.model.User
import com.petcare.repository.SessionRepository
import com.petcare.repository.UserRepository
import com.petcare.util.HashUtil
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.Optional
import java.util.UUID

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val sessionRepository: SessionRepository,
    private val passwordEncoder: PasswordEncoder
) {

    fun authenticate(email: String, password: String): Optional<User> {
        return userRepository.findByEmail(email.trim()).filter { user ->
            val hash = user.passwordHash ?: return@filter false
            if (hash.startsWith("$2") && hash.length >= 59 && passwordEncoder.matches(password, hash)) {
                true
            } else {
                val legacyMd5 = HashUtil.md5(password)
                if (hash.equals(legacyMd5, ignoreCase = true)) {
                    user.passwordHash = passwordEncoder.encode(password)
                    userRepository.save(user)
                    true
                } else {
                    false
                }
            }
        }
    }

    fun createSession(userId: Int, tokenSesion: String, ip: String?, userAgent: String?): Session {
        val s = Session()
        s.usuarioId = userId
        s.tokenSesion = tokenSesion
        s.ipAddress = ip
        s.userAgent = userAgent
        return sessionRepository.save(s)
    }

    fun createSession(userId: Int, ip: String?, userAgent: String?): Session {
        return createSession(userId, UUID.randomUUID().toString(), ip, userAgent)
    }

    fun findSessionByToken(token: String) = sessionRepository.findActiveByTokenSesion(token)
}
